tidy3d/plugins/resonance/resonance.py

```diff
@@ -0,0 +1,272 @@
+        prefactor = np.zeros((J, J), dtype=complex)
+
+        prefactor[~np.eye(J, dtype=bool)] = np.reciprocal(
+            np.subtract.outer(zvals, zvals)[~np.eye(J, dtype=bool)]
+        )
+
+        # now compute the matrices
+        u_matrices = np.zeros((3, J, J), dtype=complex)
+        for pval in range(3):
+
+            u_matrices[pval, :, :] = prefactor * (
+                np.outer(zvals, zinvl[:, : M + 1] @ signal[pval:][: M + 1])
+                + np.outer(zinvl[:, :M] @ signal[pval:][M + 1 : 2 * M + 1], zinvl[:, M])
+            )
+            u_matrices[pval, :, :] += u_matrices[pval, :, :].T
+
+            for i in range(J):
+                u_matrices[pval, i, i] = np.dot(
+                    zinvl[i, : 2 * M + 1],
+                    np.concatenate((np.arange(1, M + 2), np.arange(M, 0, -1)))
+                    * signal[pval:][: 2 * M + 1],
+                )
+
+        return u_matrices
+
+    def _normalize_eigvecs(self, eigvecs: np.ndarray[complex]) -> np.ndarray[complex]:
+        """Normalize the eigenvectors. We can approximate the Takagi decomposition
+        from the eigenvalue decomposition by just normalizing."""
+        for k in range(eigvecs.shape[1]):
+            eigvecs[:, k] /= np.sqrt(np.dot(eigvecs[:, k], eigvecs[:, k]))
+        return eigvecs
+
+    def _solve_gen_eig_prob(
+        self, a_matrix: np.ndarray[complex], b_matrix: np.ndarray[complex], rcond: float
+    ) -> Tuple[np.ndarray[complex], np.ndarray[complex]]:
+        """Solve a generalized eigenvalue problem of the form
+        Av = lambda B v. We make B nonsingular by eliminating eigenvalues
+        smaller than rcond times the largest. This method is used in
+        Michael R. Wall and Daniel Neuhauser,
+        "Extraction, through filter-diagonalization, of general quantum eigenvalues
+        or classical normal mode frequencies of a short-time segment of a signal.
+        I. Theory and application to a quantum-dynamics model,"
+        J. Chem. Phys. 102, 8001 (1995).
+
+        """
+        eigvals_b, eigvecs_b = scipy.linalg.eig(b_matrix)
+        large_inds = np.abs(eigvals_b) > rcond * np.amax(np.abs(eigvals_b))
+        eigvals_b = eigvals_b[large_inds]
+        eigvecs_b = eigvecs_b[:, large_inds]
+        eigvecs_b = self._normalize_eigvecs(eigvecs_b)
+
+        change_basis = eigvecs_b @ np.diag(1 / np.sqrt(eigvals_b))
+        new_a_matrix = change_basis.T @ a_matrix @ change_basis
+        eigvals, eigvecs_a = scipy.linalg.eig(new_a_matrix)
+        eigvecs_a = self._normalize_eigvecs(eigvecs_a)
+
+        eigvecs = change_basis @ eigvecs_a
+
+        return eigvals, eigvecs
+
+    def _find_amplitudes(
+        self,
+        signal: np.ndarray[complex],
+        prev_eigvals: np.ndarray[complex],
+        eigvals: np.ndarray[complex],
+        eigvecs: np.ndarray[complex],
+    ) -> np.ndarray[complex]:
+        """Compute the amplitudes"""
+        M = int(len(signal) / 2) - 2
```

Comment:
I think we should avoid using `M` and `J` as variable names if possible. It would be better practice to give them a concicse but descriptive names, like  `num_eigvals` for example.